### PR TITLE
Adding specific constructors to PartitionKey

### LINKED ISSUE #354: PartitionKey constructor for types
Ref: https://github.com/Azure/azure-cosmos-dotnet-v3/pull/348

Instead of using a constructor for PartitionKey as object, add one for each supported type.

### LINKED ISSUE #471: Item with null partition key throwing error on create
**Expected behavior**
Should able to create and read with null partition key

**Actual behavior**
Item with null partition key throwing error on create

**Environment summary**
SDK Version: V3

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

Partition Key currently takes an object as a parameter for the constructor. This is very confusing to users since only a few primitives are supported.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Closing issues

closes #354, fixes #471

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 1]
```diff
@@ -16,6 +18,12 @@ public sealed class PartitionKey
         /// </summary>
         public static readonly PartitionKey NonePartitionKeyValue = new PartitionKey(Documents.PartitionKey.None);
 
+        /// <summary>
+        /// The returned object represents a partition key value that allows creating and accessing documents
+        /// without a value for partition key.
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: >without a value for partition key [](start = 12, length = 33)

with null value for PK.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 2]
```diff
@@ -30,39 +38,106 @@ public sealed class PartitionKey
...
         /// Creates a new partition key value.
         /// </summary>
         /// <param name="partitionKeyValue">The value to use as partition key.</param>
-        public PartitionKey(object partitionKeyValue)
+        public PartitionKey(string partitionKeyValue)
         {
             if (partitionKeyValue == null)
             {
-                throw new ArgumentNullException(nameof(partitionKeyValue));
+                throw new ArgumentNullException($"{nameof(partitionKeyValue)} is null. Please use {nameof(PartitionKey.NullPartitionKeyValue)} to define null");
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: V3 throwing exception even for a valid PK value. 
Isn't it optimization. Then why not allow it?

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 3]
```diff
@@ -30,39 +38,106 @@ public sealed class PartitionKey
...
         /// <summary>
-        /// Gets the string representation of the partition key value.
+        /// Creates a new partition key value.
         /// </summary>
-        /// <returns>The string representation of the partition key value</returns>
-        public new string ToString()
+        /// <param name="partitionKeyValue">The value to use as partition key.</param>
+        public PartitionKey(bool partitionKeyValue)
+        {
+            this.Value = new Documents.PartitionKey(partitionKeyValue).InternalKey;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: >new Documents.PartitionKey(partitionKeyValue).InternalKey [](start = 25, length = 57)

 Just two values and choose internal value based on it.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 4]
```diff
@@ -30,39 +38,106 @@ public sealed class PartitionKey
...
+        }
+
+        /// <summary>
+        /// Convert a bool to a partition key
+        /// </summary>
+        /// <param name="partitionKeyValue">The bool value</param>
+        public static implicit operator PartitionKey(bool partitionKeyValue)
+        {
+            return new PartitionKey(partitionKeyValue);
+        }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Lets exclude implicit conversions. 
It introduces the original auto conversion of types un-knowingly.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 5]
```diff
@@ -30,39 +38,106 @@ public sealed class PartitionKey
...
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            Documents.PartitionKey docPk = value as Documents.PartitionKey;
+            if (docPk != null)
             {
-                return ((Documents.PartitionKey)this.Value).InternalKey.ToJsonString();
+                this.Value = docPk.InternalKey;
+                return;
             }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Undo if tests are green.

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/SinglePartitionRequestOptionsFactory.cs [hunk 6]
```diff
@@ -13,6 +13,6 @@ internal sealed class SinglePartitionRequestOptionsFactory : RequestOptionsFacto
     {
         public override FeedOptions CreateFeedOptions() => null;
 
-        public override PartitionKey GetPartitionKey(string itemId) => new PartitionKey(Documents.Undefined.Value);
+        public override PartitionKey GetPartitionKey(string itemId) => PartitionKey.NonePartitionKeyValue;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/SinglePartitionRequestOptionsFactory.cs]: Check with @ealsur

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs [hunk 7]
```diff
@@ -529,26 +529,32 @@ internal partial class ContainerCore : Container
             }
         }
 
-        private object CosmosElementToPartitionKeyObject(CosmosElement cosmosElement)
+        private PartitionKey CosmosElementToPartitionKeyObject(CosmosElement cosmosElement)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: Move out to new PR

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Done, it's allowed now

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Done

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs (hunk 7 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs]: It's fixed in this PR.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 8]
```diff
@@ -30,20 +40,56 @@ public sealed class PartitionKey
...
         /// </summary>
         /// <param name="partitionKeyValue">The value to use as partition key.</param>
-        public PartitionKey(object partitionKeyValue)
+        public PartitionKey(string partitionKeyValue)
         {
             if (partitionKeyValue == null)
             {
-                throw new ArgumentNullException(nameof(partitionKeyValue));
+                this.Value = PartitionKey.NullPartitionKeyInternal;
             }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: else

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 9]
```diff
@@ -52,17 +98,12 @@ public PartitionKey(object partitionKeyValue)
         /// <returns>The string representation of the partition key value</returns>
         public new string ToString()
         {
-            if (this.Value is Documents.PartitionKey)
+            if (this.Value == null)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: When will it be null?

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/SinglePartitionRequestOptionsFactory.cs (hunk 6 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/SinglePartitionRequestOptionsFactory.cs]: Makes sense taking into account the other changes

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 8 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: or return here, less nesting

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 9 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: True, it's no longer possible

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: I don't think there is a code path that leads to Value being null either

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 8 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: There is not logic after this so if else is simplest solution. No need to worry about a nesting.

**Pull request merged**